torch/serialization.py

```diff
@@ -846,7 +862,11 @@ def persistent_load(saved_id):
 
     # Load the data (which may in turn use `persistent_load` to load tensors)
     data_file = io.BytesIO(zip_file.get_record(pickle_file))
-    unpickler = pickle_module.Unpickler(data_file, **pickle_load_args)
+
+    if pickle_module.__name__ == 'pickle':
```

Comment:
This is especially fragile because this fix only works for the `pickle` module. For example, `cpickle` has a slightly different way of doing it.

I opted not to bother with the `cpickle` version because I don't see it currently used anywhere, and didn't want to add it an explicit dependency in this file. But technically, it looks like `torch.load()` lets you pass in an arbitrary module to do the unpickling, as long as it aligns with the `pickle` interface. So https://github.com/pytorch/pytorch/pull/51633 can in theory still break BC if you:
- try to `torch.load()` a tensor that was saved with a version of code that used the `torch.tensor` module
- use `torch.load()` with a custom pickling module

The combination of those two things seems unlikely (hopefully)

Reply:
Can't you do this in a "has-a" way? Your wrapper doesn't inherit from pickle.Unpickler; instead, it forwards all calls on it to the underlying pickler, except find_class which it intercepts to do our custom logic (and if it's a miss, forwards on to the underlying pickler)

Reply:
huh I might be misunderstanding- how would you do that in a "has-a" way? Wouldn't you need to explicitly write each function in the Unpickler wrapper that Unpickler.load() normally calls, and write it to forward to the underlying unpickler? It seems like we basically get all of that for free with inheritance (all calls are automatically forwarded to the base Unpickler class, except for `find_class` which we override).

Reply:
(from the [docs](https://python.readthedocs.io/en/latest/library/pickle.html#pickle.Unpickler.find_class), it sounds like overriding the class is the standard way to do it, at least for `pickle`. `cpickle` is different though)

Reply:
Oops, I said something that doesn't actually work.

Here is what actually will work though. Look at the existing unpickle implementation, you'll see that we are monkey patching the unpickler object:

```
    unpickler = pickle_module.Unpickler(f, **pickle_load_args)
    unpickler.persistent_load = persistent_load
```

You should just monkey patch `find_class` as well (and have it defer to original, pre-monkey patch `find_class` implementation if there is nothing). This will work universally for both pickle, cpickle and dill.

Reply:
Updated the PR and comment got outdated :/

> You should just monkey patch find_class as well
Ah yeah, I actually tried that. It doesn't look like `pickle` lets you do that- I get the error:
```
sts#binary,link-tree/torch/serialization.py", line 875, in _load
    unpickler.find_class = find_class
AttributeError: '_pickle.Unpickler' object attribute 'find_class' is read-only
```

So, a few things I found from looking around more:
- monkey-patching the `find_class` method works seamlessly when I use `dill`, but not for `pickle`. Subclassing appears to work for both though.
- The [stackoverflow](https://stackoverflow.com/questions/13398462/unpickling-python-objects-with-a-changed-module-path/13405732) that I based the changes on makes a distinction between `pickle` and `cpickle`, but it turns out that python3 decides which one to use for you transparently when you import `pickle` (from a stackoverflow [link](https://stackoverflow.com/questions/55406089/how-can-i-know-if-python-3-is-using-cpickle-or-pickle), which points to the implementation [here](https://github.com/python/cpython/blob/d93fbbf88e4abdd24a0a55e3ddf85b8420c62052/Lib/pickle.py#L1598-L1613). I tried to emulate using the non-c implementation by using `pickle._Unpickler` instead of `pickle.Unpickler` (from the [source](https://github.com/python/cpython/blob/d93fbbf88e4abdd24a0a55e3ddf85b8420c62052/Lib/pickle.py#L1598-L1613)), and it looks like subclassing works for both versions.
- In python2 it sounds like `pickle` and `cpickle` had a different API for mapping imports, but it looks to me like they use the same API as part of that change above. From the 

So, it looks like subclassing works in the 3 cases (dill, and both the C and python implementation of pickle). From what I can tell, `dill` and `pickle` are the only modules that we really use for serializing, so I updated the PR to always use the subclass, instead of only doing it for `pickle`. The assumption I'm making is that subclassing on `find_class` is part of the interface of any pickle serialization module- do you think that's reasonable?

I also updated the PR to not actually make the `torch.tensor` -> `torch._tensor` mapping- that needs to land along with the PR that's actually changing the module.